reactive/calico.py

```diff
@@ -21,19 +26,96 @@
 ETCD_CERT_PATH = os.path.join(CALICOCTL_PATH, 'etcd-cert')
 ETCD_CA_PATH = os.path.join(CALICOCTL_PATH, 'etcd-ca')
 CALICO_CIDR = '192.168.0.0/16'
+CALICO_UPGRADE_DIR = '/opt/calico-upgrade'
 
 
 @hook('upgrade-charm')
 def upgrade_charm():
     remove_state('calico.binaries.installed')
     remove_state('calico.cni.configured')
+    remove_state('calico.service.installed')
+    remove_state('calico.pool.configured')
+    remove_state('calico.cni.configured')
+    remove_state('calico.npc.deployed')
     try:
         log('Deleting /etc/cni/net.d/10-calico.conf')
         os.remove('/etc/cni/net.d/10-calico.conf')
     except FileNotFoundError as e:
         log(e)
+    if is_leader() and not leader_get('calico-v3-data-ready'):
+        leader_set({
+            'calico-v3-data-migration-needed': True,
+            'calico-v3-npc-cleanup-needed': True,
+            'calico-v3-completion-needed': True
+        })
+
+
+@when('leadership.is_leader', 'leadership.set.calico-v3-data-migration-needed',
+      'etcd.available', 'calico.etcd-credentials.installed')
+def upgrade_v3_migrate_data():
+    status_set('maintenance', 'Migrating data to Calico 3')
+    try:
+        calico_upgrade.configure()
+        calico_upgrade.dry_run()
+        calico_upgrade.start()
+    except Exception:
+        log(traceback.format_exc())
+        message = 'Calico upgrade failed, see debug log'
+        status_set('blocked', message)
+        return
+    leader_set({'calico-v3-data-migration-needed': None})
+
+
+@when('leadership.is_leader')
+@when_not('leadership.set.calico-v3-data-migration-needed')
+def v3_data_ready():
+    leader_set({'calico-v3-data-ready': True})
```

Comment:
We only ever seem to set this to true. Do we need this data?

Reply:
Yes. It is `None` by default. It is used here to determine whether or not an upgrade is needed:

https://github.com/charmed-kubernetes/layer-calico/blob/c4ec07666ab118d84f1c99986b4024815a4f6cc3/reactive/calico.py#L45

Reply:
I guess I'm getting tripped up on the naming. It isn't obvious that this indicated that an upgrade is needed or that it implies we don't need a data migration. Unfortunately, I'm drawing a blank for helpful suggestions.

Reply:
Some alternative names:

calico-v3-data-available
calico-v3-components-runnable

Another option is to remove `calico-v3-data-ready`, and invert the truthiness of `calico-v3-data-migration-needed` (so it becomes something like `calico-v3-data-migration-complete`) so followers can wait for that instead. What do you think?